File: go/leetcode/278-first_bad_version.go
/* First Bad Version
Source		: https://leetcode.com/problems/first-bad-version/
Level		: Easy
Description	:

Example 1:
Input: n = 5, bad = 4
Output: 4
Explanation:
call isBadVersion(3) -> false
call isBadVersion(5) -> true
call isBadVersion(4) -> true
Then 4 is the first bad version.

Example 2:
Input: n = 1, bad = 1
Output: 1

Solution source: https://www.programiz.com/dsa/binary-search
*/

package leetcode

// FirstBadVersion implements binary tree to find the first bad version
func (soln Solution) FirstBadVersion(n int) int {
	// list of version
	low, high, mid := 1, n, 0
	firstBad := n

	for low <= high {
		mid = (low + high) / 2
		switch {
		case isBadVersion(mid):
			if mid < firstBad {
				firstBad = mid
			}
			high = mid - 1
		default:
			low = mid + 1
		}
	}

	return firstBad
}

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */
func isBadVersion(version int) bool {
	// this function is not implemented here
	return true
}

File: leetcode/pysolution/278_first_bad_version.py
"""
LeetCode Problem : First Bad Version
Topic            : Binary Search, Interactive
Level            : Easy
URL              : https://leetcode.com/problems/first-bad-version/description
"""


# The isBadVersion API is already defined for you.
def isBadVersion(version: int) -> bool:
    return version >= 4


class Solution:
    def firstBadVersion(self, n: int) -> int:
        """
        Find first version where code broken by using binary search approach.
        Assuming the version order goes from 1 to n.
        Args:
            n: int representing the total number of versions
        Returns:
            int: the first bad version number
        Examples:
            n = 5, bad = 4 -> output = 4
            All versions from 4 onwards are bad
        Solutions:
            Binary Search
        Time Complexity:
            O(log n) - binary search halves search space each iteration
        Space Complexity:
            O(1) - only uses a constant amount of extra space
        """

        low, high = 1, n

        while low < high:
            mid = low + (high - low) // 2

            if isBadVersion(mid):
                high = mid
            else:
                low = mid + 1

        return low


def run_tests():
    inputs = {"case_1": [5]}
    outputs = {"case_1": 4}

    solution = Solution()

    for case, input in inputs.items():
        result = solution.firstBadVersion(input[0])
        assert result == outputs[case], f"{case}: expected {outputs[case]}, got {result}"

    print("✅ All tests passed!")


if __name__ == "__main__":
    run_tests()
